File: engine/src/main/java/com/terralite/engine/entity/Entity.java
package com.terralite.engine.entity;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

public final class Entity {
    private final EntityId id;
    private final Map<ComponentType<?>, Object> components = new LinkedHashMap<>();

    public Entity(EntityId id) {
        this.id = Objects.requireNonNull(id, "id");
    }

    public EntityId id() {
        return id;
    }

    public <T> Entity set(ComponentType<T> type, T component) {
        Objects.requireNonNull(type, "type");
        components.put(type, type.type().cast(Objects.requireNonNull(component, "component")));
        return this;
    }

    public <T> Optional<T> get(ComponentType<T> type) {
        Objects.requireNonNull(type, "type");
        return Optional.ofNullable(type.type().cast(components.get(type)));
    }

    public <T> T require(ComponentType<T> type) {
        return get(type).orElseThrow(() -> new IllegalArgumentException("Missing component: " + type.name()));
    }

    public boolean has(ComponentType<?> type) {
        return components.containsKey(Objects.requireNonNull(type, "type"));
    }

    public <T> Optional<T> remove(ComponentType<T> type) {
        Objects.requireNonNull(type, "type");
        return Optional.ofNullable(type.type().cast(components.remove(type)));
    }

    public int componentCount() {
        return components.size();
    }
}

File: engine/src/commonMain/kotlin/com/terralite/engine/entity/Entity.kt
package com.terralite.engine.entity

class Entity(@get:JvmName("id") val id: EntityId) {
    private val components: MutableMap<ComponentType<*>, Any> = LinkedHashMap()

    fun <T : Any> set(type: ComponentType<T>, component: T): Entity {
        components[type] = component
        return this
    }

    @Suppress("UNCHECKED_CAST")
    fun <T : Any> get(type: ComponentType<T>): T? = components[type] as T?

    fun <T : Any> require(type: ComponentType<T>): T =
        get(type) ?: throw IllegalArgumentException("Missing component: ${type.name}")

    fun has(type: ComponentType<*>): Boolean = components.containsKey(type)

    @Suppress("UNCHECKED_CAST")
    fun <T : Any> remove(type: ComponentType<T>): T? = components.remove(type) as T?

    fun componentCount(): Int = components.size
}
